connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Flatten.java

```diff
@@ -0,0 +1,237 @@
+        return newRecord(record, null, newValue);
+    }
+
+    private void applySchemaless(Map<String, Object> originalRecord, String fieldNamePrefix, Map<String, Object> newRecord) {
+        for (Map.Entry<String, Object> entry : originalRecord.entrySet()) {
+            final String fieldName = fieldName(fieldNamePrefix, entry.getKey());
+            Object value = entry.getValue();
+            if (value == null) {
+                newRecord.put(fieldName(fieldNamePrefix, entry.getKey()), null);
+                return;
+            }
+            switch (ConnectSchema.schemaType(value.getClass())) {
+                case INT8:
+                case INT16:
+                case INT32:
+                case INT64:
+                case FLOAT32:
+                case FLOAT64:
+                case BOOLEAN:
+                case STRING:
+                case BYTES:
+                    newRecord.put(fieldName(fieldNamePrefix, entry.getKey()), entry.getValue());
+                    break;
+                case MAP:
+                    final Map<String, Object> fieldValue = requireMap(entry.getValue(), PURPOSE);
+                    applySchemaless(fieldValue, fieldName, newRecord);
+                    break;
+                default:
+                    throw new ConnectException("Flatten transformation does not support " + entry.getValue().getClass()
+                            + " for record without schemas (for field " + fieldName + ").");
+            }
+        }
+    }
+
+    private R applyWithSchema(R record) {
+        final Struct value = requireStruct(operatingValue(record), PURPOSE);
+
+        Schema updatedSchema = schemaUpdateCache.get(value.schema());
+        if (updatedSchema == null) {
+            final SchemaBuilder builder = SchemaUtil.copySchemaBasics(value.schema(), SchemaBuilder.struct());
+            buildUpdatedSchema(value.schema(), "", builder);
+            updatedSchema = builder.build();
+            schemaUpdateCache.put(value.schema(), updatedSchema);
+        }
+
+        final Struct updatedValue = new Struct(updatedSchema);
+        buildWithSchema(value, "", updatedValue);
+        return newRecord(record, updatedSchema, updatedValue);
+
+    }
+
+    private void buildUpdatedSchema(Schema schema, String fieldNamePrefix, SchemaBuilder newSchema) {
+        for (Field field : schema.fields()) {
+            final String fieldName = fieldName(fieldNamePrefix, field.name());
+            switch (field.schema().type()) {
+                case INT8:
+                case INT16:
+                case INT32:
+                case INT64:
+                case FLOAT32:
+                case FLOAT64:
+                case BOOLEAN:
+                case STRING:
+                case BYTES:
+                    newSchema.field(fieldName, field.schema());
+                    break;
+                case STRUCT:
+                    buildUpdatedSchema(field.schema(), fieldName, newSchema);
```

Comment:
That was a really good observation. Turns out this isn't too hard to layer onto the existing implementation, although I have to admit that I would think the semantics can be confusing to reason about. Thankfully I don't think most folks use default values on anything but individual primitive fields.